File: GameServer/Physics/Physic.cs
using System;
using GameServer.Models;

namespace GameServer.Physics
{
    public class Physic
    {
        public static Vector2 GetLeftParallelVectorToIntersectionNormal(Vector2 movementVector, double intersectionDistance, Vector2 intersectionNormal)
        {
            var movementVectorLength = movementVector.Length();
            var realMovementDirectionVector = Vector2.Normalize(movementVector) * intersectionDistance;

            var leftMovementDirectionVector = movementVector - realMovementDirectionVector;
            return GetParallelVectorToNormal(leftMovementDirectionVector, intersectionNormal);
        }

        public static Vector2 GetParallelVectorToNormal(Vector2 vector, Vector2 normal)
        {
            return ProjectVector(vector, RotateVector(normal, (float)(Math.PI / 2)));
        }

        public static Vector2 ProjectVector(Vector2 vector, Vector2 projectionVector)
        {
            return projectionVector.Normalize() * Vector2.Dot(vector, projectionVector) / projectionVector.Length();
        }
        
        public static Vector2 RotateVector(Vector2 v, double angle)
        {
            var ca = Math.Cos(angle);
            var sa = Math.Sin(angle);
            return new Vector2(ca * v.X - sa * v.Y, sa * v.X + ca * v.Y);
        }
    }
}

File: GameServer/wwwroot/js/Physics/Physic.js
const Vector2 = require('./Vector2.js');

function Physic() {

};

Physic.GetLeftParallelVectorToIntersectionNormal = function(movementVector, intersectionDistance, intersectionNormal) {

    const realMovementDirectionVector = Vector2.Multiply(Vector2.Normalize(movementVector), intersectionDistance);
    const leftMovementDirectionVector = Vector2.Subtract(movementVector, realMovementDirectionVector);

    return Physic.GetParallelVectorToNormal(leftMovementDirectionVector, intersectionNormal);
};

Physic.GetParallelVectorToNormal = function (vector, normalVector) {
    return Physic.ProjectVector(vector, Physic.RotateVector(normalVector, Math.PI / 2.0));
};

Physic.ProjectVector = function (vector, projectionVector) {

    const normalizedProjectionVector = projectionVector.Normalize();
    const dotProduct = Vector2.Dot(vector, projectionVector);

    return Vector2.Multiply(normalizedProjectionVector, dotProduct / projectionVector.Length());
};

Physic.RotateVector = function (vector, angle) {

    const cosAngle = Math.cos(angle);
    const sinAngle = Math.sin(angle);

    const x = cosAngle * vector.GetX() - sinAngle * vector.GetY();
    const y = sinAngle * vector.GetX() + cosAngle * vector.GetY();
    return new Vector2(x, y);
};

// Export module
if (typeof module !== 'undefined' && module.hasOwnProperty('exports')) {
    module.exports = Physic;
}
